storage/innobase/handler/ha_innodb.cc

```diff
@@ -17116,6 +17121,81 @@ int innobase_rollback_by_xid(handlerton* hton, XID* xid)
 	}
 }
 
+#ifndef EMBEDDED_LIBRARY
+/**
+  This function is used to rollback one X/Open XA distributed transaction
+  which is in the prepared state asynchronously.
+
+  It only set the transaction's status to ACTIVE and persist the status.
+  The transaction will be rolled back by background rollback thread.
+
+  @param xid X/Open XA transaction identification
+
+  @return 0 or error number
+*/
+int innobase_recover_rollback_by_xid(const XID *xid)
+{
+  DBUG_EXECUTE_IF("innobase_xa_fail", return XAER_RMFAIL;);
+
+  if (high_level_read_only)
+    return XAER_RMFAIL;
+
+  /*
+    trx_get_trx_by_xid() sets trx's xid to null. Thus only one call for any
+    given XID can find the transaction. Subsequent calls by other threads
+    would return nullptr. That is what guarantees that no other thread can be
+    modifying the state of the transaction at this point.
+  */
+  trx_t *trx= trx_get_trx_by_xid(xid);
+  if (!trx)
+    return XAER_RMFAIL;
+
+  // ddl should not be rolled back through recovery
+  ut_ad(!trx->dict_operation);
+  ut_ad(trx->is_recovered);
+  ut_ad(trx->state == TRX_STATE_PREPARED);
+
+#ifdef WITH_WSREP
+  ut_ad(!wsrep_is_wsrep_xid(&trx->xid));
+#endif
+
+  if (trx->rsegs.m_redo.undo)
+  {
+    ut_ad(trx->rsegs.m_redo.undo->rseg == trx->rsegs.m_redo.rseg);
+
+    /* This change will be made durable in innobase_tc_log_recovery_done(). */
+    mtr_t mtr;
+    mtr.start();
+    trx_undo_set_state_at_prepare(trx, trx->rsegs.m_redo.undo, true, &mtr);
+    mtr.commit();
+
+    ut_ad(mtr.commit_lsn() > 0);
+  }
+
+  /* Change the trx state to TRX_STATE_ACTIVE. Background rollback thread
+     will rollback this trx later. */
+  trx->state= TRX_STATE_ACTIVE;
```

Comment:
In InnoDB, we usually don’t indent the second line of a comment deeper than the first line.

We could mention the actual function name here and omit the kind-of redundant first part:
```c++
  /* trx_rollback_recovered() will roll back this later. */
  trx->state= TRX_STATE_ACTIVE;
```
We could also remove the previous comment about durability and do it here:
```c++
  /* The above state change from XA PREPARE will be made durable in
  innobase_tc_log_recovery_done(), which will also initiate
  trx_rollback_recovered() to roll back this transaction. */
  trx->state= TRX_STATE_ACTIVE;
```

Reply:
fixed